File: Leetcode 3601-3700/3624/3624.java
// Leetcode 3624: Number of Integers With Popcount-Depth Equal to K II
// https://leetcode.com/problems/number-of-integers-with-popcount-depth-equal-to-k-ii/
// Solved on 19th of December, 2025
class Solution {
    /**
     * Calculates the number of integers with a specific popcount-depth within given ranges,
     * and supports updates to the array of numbers.
     * @param nums An array of long integers.
     * @param queries A 2D array of queries. Each query can be of type 1 (range query) or type 2 (update query).
     * @return An array of integers, where each element is the result of a type 1 query.
     */
    public int[] popcountDepth(long[] nums, long[][] queries) {
        int n = nums.length;
        int[] preDepth = new int[65];
        preDepth[1] = 0;
        for (int i = 2; i <= 64; i++) {
            preDepth[i] = 1 + preDepth[Integer.bitCount(i)];
        }

        int[] numDepths = new int[n];
        int[][] bit = new int[n + 1][6];

        for (int i = 0; i < n; i++) {
            long val = nums[i];
            int d = 0;
            if (val > 1) {
                d = 1 + preDepth[Long.bitCount(val)];
            }
            numDepths[i] = d;
            add(bit, i + 1, d, 1);
        }

        int queryCount = 0;
        for (long[] q : queries) {
            if (q[0] == 1) {
                queryCount++;
            }
        }

        int[] result = new int[queryCount];
        int resIdx = 0;

        for (long[] q : queries) {
            int type = (int) q[0];
            if (type == 1) {
                int l = (int) q[1];
                int r = (int) q[2];
                int k = (int) q[3];
                if (k > 5) {
                    result[resIdx++] = 0;
                } else {
                    result[resIdx++] = query(bit, r + 1, k) - query(bit, l, k);
                }
            } else {
                int idx = (int) q[1];
                long val = q[2];
                int oldD = numDepths[idx];
                int newD = 0;
                if (val > 1) {
                    newD = 1 + preDepth[Long.bitCount(val)];
                }
                if (oldD != newD) {
                    add(bit, idx + 1, oldD, -1);
                    add(bit, idx + 1, newD, 1);
                    numDepths[idx] = newD;
                }
            }
        }
        return result;
    }

    private void add(int[][] bit, int index, int k, int delta) {
        while (index < bit.length) {
            bit[index][k] += delta;
            index += index & -index;
        }
    }

    private int query(int[][] bit, int index, int k) {
        int sum = 0;
        while (index > 0) {
            sum += bit[index][k];
            index -= index & -index;
        }
        return sum;
    }
}

File: Leetcode 3601-3700/3624/3624.py
# Leetcode 3624: Number of Integers With Popcount-Depth Equal to K II
# https://leetcode.com/problems/number-of-integers-with-popcount-depth-equal-to-k-ii/
# Solved on 19th of December, 2025
class Solution:
    def popcountDepth(self, nums: list[int], queries: list[list[int]]) -> list[int]:
        """
        Calculates the number of integers with a specific popcount-depth within given ranges and handles updates.

        Args:
            nums: A list of integers.
            queries: A list of queries. Each query can be either a type 1 query [1, l, r, k] or a type 2 query [2, idx, val].
        Returns:
            A list of integers, where each element is the answer to a type 1 query.
        """
        n = len(nums)
        maxDepth = 5
        trees = [[0] * (n + 1) for _ in range(maxDepth + 1)]
        currentDepths = [0] * n

        def getDepth(num):
            depth = 0
            while num != 1:
                num = bin(num).count('1')
                depth += 1
            return depth

        def update(treeIdx, idx, delta):
            idx += 1
            while idx <= n:
                trees[treeIdx][idx] += delta
                idx += idx & (-idx)

        def query(treeIdx, idx):
            idx += 1
            total = 0
            while idx > 0:
                total += trees[treeIdx][idx]
                idx -= idx & (-idx)
            return total

        for i in range(n):
            d = getDepth(nums[i])
            currentDepths[i] = d
            if d <= maxDepth:
                update(d, i, 1)

        ans = []
        for q in queries:
            if q[0] == 1:
                l, r, k = q[1], q[2], q[3]
                if k <= maxDepth:
                    count = query(k, r) - query(k, l - 1)
                    ans.append(count)
                else:
                    ans.append(0)
            else:
                idx, val = q[1], q[2]
                oldDepth = currentDepths[idx]
                newDepth = getDepth(val)

                if oldDepth == newDepth:
                    nums[idx] = val
                    continue

                if oldDepth <= maxDepth:
                    update(oldDepth, idx, -1)
                if newDepth <= maxDepth:
                    update(newDepth, idx, 1)

                currentDepths[idx] = newDepth
                nums[idx] = val

        return ans